pylearn2/datasets/hdf5.py

```diff
@@ -233,7 +239,26 @@ def next(self):
             try:
                 this_data = data[next_index]
             except TypeError:
-                this_data = data[next_index, :]
+                # FB: Why this try..except is there? I think this is useless.
+                # Do not hide the original if we can't fall back.
+                # FV: This is triggered if the shape of next_index is
+                # incompatible with the shape of the dataset. See for an
+                # example test_hdf5_topo_view(), where where i
+                # next.index.shape = (10,) and data is 'data': <HDF5
+                # dataset "y": shape (10, 3), type "<f8">
+                # I think it would be better to explicitly check if
+                # next_index.shape is incompatible with data.shape, for
+                # instance checking if next_index.ndim == data.ndim
+                if data.ndim > 1:
+                    this_data = data[next_index, :]
+                else:
+                    raise
```

Comment:
OK, I just tried it with h5py.
The problem is that, when `next_index` is an `ndarray` and not a list, `data[next_index]` is interpreted as indexing by a mask, not by a list of indices, and it fails with the following error:
```
TypeError: PointSelection __getitem__ only works with bool arrays
```
The solution is indeed to force the interpretation of `next_index` as an index in only one dimension, but a more robust way of doing that is to always call `data[next_index, ...]`, which works even if `data` has only one dimension.
Could you update the comment again, remove the try/catch and replace it by `data[next_index, ...]`?
If my comment is not clear enough, let me know and I'll submit a PR to yours.